models/src/main/scala/coop/rchain/models/rholang/implicits.scala

```diff
@@ -502,4 +502,131 @@ object implicits {
+            }
+          case EVarBody(_)                => true
+          case EEvalBody(chan)            => ChannelIsForgeable(chan)
+          case ENotBody(ENot(p))          => ParIsForgeable(p)
+          case ENegBody(ENeg(p))          => ParIsForgeable(p)
+          case EMultBody(EMult(p1, p2))   => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EDivBody(EDiv(p1, p2))     => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EPlusBody(EPlus(p1, p2))   => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EMinusBody(EMinus(p1, p2)) => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case ELtBody(ELt(p1, p2))       => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case ELteBody(ELte(p1, p2))     => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EGtBody(EGt(p1, p2))       => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EGteBody(EGte(p1, p2))     => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EEqBody(EEq(p1, p2))       => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case ENeqBody(ENeq(p1, p2))     => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EAndBody(EAnd(p1, p2))     => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EOrBody(EOr(p1, p2))       => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EMethodBody(e) =>
+            ParIsForgeable(e.target) && e.arguments.forall(ParIsForgeable.apply)
+          case EMatchesBody(EMatches(target, pattern)) =>
+            ParIsForgeable(target) && ParIsForgeable(pattern)
+          case EPercentPercentBody(EPercentPercent(p1, p2)) =>
+            ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EPlusPlusBody(EPlusPlus(p1, p2))     => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case EMinusMinusBody(EMinusMinus(p1, p2)) => ParIsForgeable(p1) && ParIsForgeable(p2)
+          case ExprInstance.Empty                   => true
+        }
+    }
+  implicit val MatchIsForgeable: IsForgeable[Match] =
+    new IsForgeable[Match] {
+      def isForgeable(mat: Match) =
+        ParIsForgeable(mat.target) && mat.cases.forall(MatchCaseIsForgeable.apply)
+    }
+
+  implicit val MatchCaseIsForgeable: IsForgeable[MatchCase] =
+    new IsForgeable[MatchCase] {
+      def isForgeable(mc: MatchCase) =
+        ParIsForgeable(mc.pattern) && ParIsForgeable(mc.source)
+    }
+
+  implicit val BundleIsForgeable: IsForgeable[Bundle] =
+    new IsForgeable[Bundle] {
+      def isForgeable(bundle: Bundle) = ParIsForgeable(bundle.body)
+    }
+
+  implicit val ConnectiveIsForgeable: IsForgeable[Connective] =
+    new IsForgeable[Connective] {
+      def isForgeable(conn: Connective) =
+        conn.connectiveInstance match {
+          case ConnAndBody(body)        => body.ps.forall(ParIsForgeable.apply)
+          case ConnOrBody(body)         => body.ps.forall(ParIsForgeable.apply)
+          case ConnNotBody(body)        => ParIsForgeable(body)
+          case VarRefBody(_)            => true
```

Comment:
OK, so I see that you took the approach that any variable is forgeable, even if it might reference unforgeable one?

Reply:
We do the check post substitution. So any variables left are forgeable ones.